### DISCUSSION TITLE
Modeling a Hollow Tube with Tangent Radial Arcs in CadQuery: Seeking Guidance

### DISCUSSION THREAD

**Discussion by Obi-Wan-Kenometrics** [CadQuery/cadquery#1415] [Q&A]
Hello, I have experience with Creo, a robust CAD software, and am now delving into CadQuery to explore specifying geometry through code. In Creo, I would solve my current problem by offsetting points from a coordinate system. Specifically, I input the X, Y, and Z coordinates for five points, create a datum curve connecting these points with straight lines, specify the radii for each corner, and finally perform a sweep with a hollow tube cross-section. Now, as I transition into a code-based geometry specification in CadQuery, I am unsure of the recommended approach to model this geometry. I appreciate any insights or suggestions from the community here.

Problem Description:
I aim to model a hollow tube with multiple radial arcs tangent to the straight arm segments. Below are the specifics of the geometry:

**Inner Diameter**: Ø25.4mm
**Wall Thickness**: 5.00mm
**Radii at Points**:
-         Point 2: 25.4mm
-         Point 3: 25.4mm
-         Point 4: 50.8mm

**X, Y, Z Coordinates defining the centerline of the path**:

-     Point 1: 0, 0, 0mm
-     Point 2: 75, -30, 60mm
-     Point 3: 110, 25, 0mm
-     Point 4: 165, -50, -45mm
-     Point 5: 220, -85, 90mm

![Hose_Example](https://github.com/CadQuery/cadquery/assets/147356555/b367e6b3-03b7-4bb2-8ffe-074a6580fdca)

I am new to CadQuery and am seeking guidance on how to approach this modeling task, especially in translating the method I used in Creo to CadQuery.

**Comment by adam-urbanczyk** [comment]
Fillet for 3D wires is not implemented, so you'll need to build the spine by hand calling e.g. `Edge.makeLine` and `Edge.makeTangentArc`

**Comment by arcilyes** [comment]
Hi Obi, I have tried drawing your case using the occt API in Cadquery. I hope this helps.

import cadquery as cq
from OCP.gp import gp_Pnt, gp_Vec, gp_Dir, gp_Pln
from OCP.ChFi2d import ChFi2d_AnaFilletAlgo
from OCP.BRepBuilderAPI import BRepBuilderAPI_MakeEdge, BRepBuilderAPI_MakeWire
from cadquery.occ_impl.shapes import Wire

point1 = gp_Pnt(0.0, 0.0, 0.0)
point2 = gp_Pnt(75.0, -30.0, 60.0)
point3 = gp_Pnt(110.0, 25.0, 0.0)
point4 = gp_Pnt(165.0, -50.0, -45.0)
point5 = gp_Pnt(220.0, -85.0, 90.0)

edge1 = BRepBuilderAPI_MakeEdge(point1, point2).Edge()
edge2 = BRepBuilderAPI_MakeEdge(point2, point3).Edge()
edge3 = BRepBuilderAPI_MakeEdge(point3, point4).Edge()
edge4 = BRepBuilderAPI_MakeEdge(point4, point5).Edge()

vec1 = gp_Vec(point1, point2)
vec2 = gp_Vec(point2, point3)
vec3 = gp_Vec(point3, point4)
vec4 = gp_Vec(point4, point5)

dir1 = gp_Dir(vec1.Crossed(vec2))
plane1 = gp_Pln(point1, dir1)
f = ChFi2d_AnaFilletAlgo()
f.Init(edge1, edge2, plane1)
f.Perform(25.4)
fillet2 = f.Result(edge1, edge2)

dir2 = gp_Dir(vec2.Crossed(vec3))
plane2 = gp_Pln(point2, dir2)
f = ChFi2d_AnaFilletAlgo()
f.Init(edge2, edge3, plane2)
f.Perform(25.4)
fillet3 = f.Result(edge2, edge3)

dir3 = gp_Dir(vec3.Crossed(vec4))
plane3 = gp_Pln(point3, dir3)
f = ChFi2d_AnaFilletAlgo()
f.Init(edge3, edge4, plane3)
f.Perform(50.8)
fillet4 = f.Result(edge3, edge4)

wire_builder = BRepBuilderAPI_MakeWire()
wire_builder.Add(edge1)
wire_builder.Add(fillet2)
wire_builder.Add(edge2)
wire_builder.Add(fillet3)
wire_builder.Add(edge3)
wire_builder.Add(fillet4)
wire_builder.Add(edge4)

path_wire = Wire(wire_builder.Wire())

OCCT_plane = cq.Plane(point1, normal=(vec1.X(), vec1.Y(), vec1.Z()))

outer_wire = cq.Workplane(OCCT_plane).circle(20.4)
inner_wire = cq.Workplane(OCCT_plane).circle(15.4)

solid = outer_wire.sweep(path_wire)
solid = solid.cut(inner_wire.sweep(path_wire))

show_object(solid)
![CadqueryPipe](https://github.com/CadQuery/cadquery/assets/68187936/5dc97d62-4bb5-412b-870c-d9d9415b7f69)
